Add check for the MADV_FREE/fork arm64 Linux kernel bug

prekageo opened the pull request:
Older arm64 Linux kernels have a bug that could lead to data corruption during
background save under the following scenario:

1) jemalloc uses MADV_FREE on a page,
2) jemalloc reuses and writes the page,
3) Redis forks the background save process, and
4) Linux performs page reclamation.

Under these conditions, Linux will reclaim the page wrongfully and the
background save process will read zeros when it tries to read the page.

The bug has been fixed in Linux with commit:
ff1712f953e27f0b0718762ec17d0adb15c9fd0b ("arm64: pgtable: Ensure dirty bit is
preserved across pte_wrprotect()")

prekageo commented:
Related to issue #8124

oranagra commented on src/server.c:
```diff
@@ -4997,6 +5001,88 @@ void linuxMemoryWarnings(void) {
     }
 }
+
+static int smapsGetSharedDirty(pid_t pid, unsigned long addr) {
```
please add a top comment that stats the input and output.

oranagra commented on src/server.c:
```diff
@@ -4997,6 +5001,88 @@ void linuxMemoryWarnings(void) {
+    return val;
+}
+
+void linuxMadvFreeForkBugCheck(void) {
```
please add a big comment explaining:
1. what it does (i.e. check for kernel COW bug, basically your commit comment).
2. why we do it (i.e. we want to warn the user and exit)
3. how we do it (what are the steps, or alternatively just add a few comments inside the function)

oranagra commented on src/server.c:
```diff
@@ -4997,6 +5001,88 @@ void linuxMemoryWarnings(void) {
+
+        exit(0);
+    } else {
+        if (!smapsGetSharedDirty(pid, (unsigned long)p))
```
are we sure the parent can always read the child's smaps?
@yossigo care to comment?

i was thinking we may need to have the child read it and communicate the result to the parent via the pipe.

yossigo replied:
IIRC, we won't be able to read smaps of a child if privileged (e.g. running with a Posix capability, etc.).

oranagra commented on src/server.c:
```diff
@@ -4997,6 +5001,88 @@ void linuxMemoryWarnings(void) {
+    serverAssert(pid >= 0);
+
+    if (!pid) {
+        ret = read(pipefd[0], &buf, 1);
```
so this is just used to keep the child alive until the parent is done looking into it?

oranagra commented on src/server.c:
```diff
@@ -4997,6 +5001,88 @@ void linuxMemoryWarnings(void) {
         serverLog(LL_WARNING,"WARNING you have Transparent Huge Pages (THP) support enabled in your kernel. This will create latency and memory usage issues with Redis. To fix this issue run the command 'echo madvise > /sys/kernel/mm/transparent_hugepage/enabled' as root, and add it to your /etc/rc.local in order to retain the setting after a reboot. Redis must be restarted after THP is disabled  ...
     }
 }
+
```
let's put all this code inside an ifdef that'll include it only on arm64 build (and linux, which it already is).

oranagra commented on src/server.c:
```diff
@@ -4997,6 +5001,88 @@ void linuxMemoryWarnings(void) {
+        exit(0);
+    } else {
+        if (!smapsGetSharedDirty(pid, (unsigned long)p))
+            serverLog(LL_WARNING,"WARNING Your kernel has a bug that could lead to data corruption during background save. Please upgrade to the latest stable kernel.");
```
i think we want to `exit(1)` here, people might not notice the warning.

oranagra reviewed (commented):
@prekageo thanks a lot for this PR.
requested a few small changes (mainly comments).

yossigo commented:
I'm not a fan of going above and beyond to handle environmental issues, but I think in this case it is justified due to the risk of irrecoverable corruption on affected hardware/kernel combinations. Taking the extra step of actually aborting seems even more extreme, and IMHO should either be avoided or softened by:

1. Having a configuration parameter to disable that, should users accept the risk.
2. When this gets ported to 6.0.x, make sure by default we just produce a warning but don't exit - not to introduce regressions.

oranagra commented:
@prekageo i pushed a commit to your branch with the changes Yossi and myself asked for.
Please take a look and let me know if something looks wrong, and please test it for me.
With the current code redis should by default exit on startup on the affected platform, and run successfully on the fixed kernel.
running it with `src/redis-server --ignore-warnings ARM64-COW-BUG` should let it run despite the warning.

@redis/core-team please approve a new generic `ignore-warnings` config. by default the warning is not ignored, and redis will exit on startup. but when cherry picking this commit to 6.0, we'll put a default value into config.c which will by default ignore the warning.

yossigo commented:
@oranagra Do we completely give up the idea of disabling muzzy decay on arm64 and purely relying on this check?

oranagra commented:
@yossigo i don't know the full implications of disabling it, and i don't want to blindly disable it for all arm64 builds.
we can suggest in the warning message that disabling muzzy decay is an option if you can't upgrade your kernel.
or add a reference to this PR (and put that info in the top comment).

what we can maybe also do is to check at runtime if jemmalloc has muzzy decay enabled or not, and if disabled avoid the warning (don't even bother to run the fork check). but considering it's so reliable, and a system with this bug really does need an upgrade IMHO, i don't think i wanna skip that check if muzzy is already disabled (users can disable the check with the new config if they want)

so what do you think? is improving the warning message enough?
care to suggest a wording?

prekageo commented:
I've fixed a couple of things and pushed some commits.

Most importantly, there was a case where Linux will merge adjacent VMAs and the check will fail. I've tried to fix that.

Since we are introducing `#ifdef __arm64__` code regions, should we add some CI for arm64 as well?

I've rebased all commits on top of latest unstable but I didn't modify them. Do you want me to fixup the original commits? Let me know what is your preference regarding Git history.

oranagra commented:
Thanks @prekageo it all looks great to me (I assume you tested it). 

Regarding git, I prefer to merge it using squash, so it'll appear in Redis history as one commit. 
I can use something form the top comment of the PR or anything else you wanna provide as the commit comment. 

I think the only thing missing is for Yossi to respond what he wants to do about muzzy decay. I think that documenting it in the warning or referring to github is enough.

prekageo commented:
I've tested it on an AWS EC2 m6g instance. I've used QEMU to run a custom compiled kernel with and without the fix and launched redis-server. I've also tested the ` --ignore-warnings ARM64-COW-BUG` parameter.

Regarding git, feel free to reword the message as you like.

itamarhaber reviewed (dismissed):
LGTM but I didn't dive into the issue or the resolution

yossigo commented:
@oranagra I was under the wrong impression it's a confirmed option - if we're not sure about the implications I think we should try to keep things simple and just promote kernel upgrades.

oranagra commented:
@yossigo we confirmed that it solves the problem. personally i was too lazy to look into it myself and study the implications.
but we can mention it in the comment and whoever uses it can decide to look into it.

madolson commented on src/server.c:
```diff
@@ -5054,6 +5058,21 @@ void monitorCommand(client *c) {
 
 /* =================================== Main! ================================ */
 
+int CheckIgnoreWarning(const char *warning) {
```
nitpick: checkIgnoreWarning(), we can fix this when we merge though.

oranagra commented:
one thing to note about the config API, is that it doesn't allow incremental change.
i.e. if the default in the conf file is `ignore-warnings some-warning1` and user wants to add another, he'll need to do `--ignore-warnings "some-warning1 some-warning2"`.
then if the default in the config file changes, the user override wins (won't benefit from the default being changed).

maybe we could have done something similar to the `save` config or something more complex.
just putting this out there so you can respond if you think this is a problem (considering we're trying to create a general warnings mechanism).

@yossigo ^^

yossigo commented:
@oranagra I think the `save` special handling was troublesome enough to justify not repeating that pattern, so I'm willing to take this risk. Besides, I don't expect the default value of this parameter to be anything but null (why would we add a warning if we disable it by default)?
